app/services/discourse_subscriptions/campaign.rb

```diff
@@ -17,26 +17,39 @@ def refresh_data
         product_ids = product_ids.include?(campaign_product) ? [campaign_product] : []
       end
 
-      amount = 0.00
-      subscriptions = get_subscription_data
-      subscriptions = filter_to_subscriptions_products(subscriptions, product_ids)
+      begin
+        amount = 0.00
+        subscriptions = get_subscription_data
+        subscriptions = filter_to_subscriptions_products(subscriptions, product_ids)
 
-      # Fetch product purchases
-      one_time_payments = get_one_time_payments(product_ids)
-      one_time_payments.each { |c| amount += c[:price].to_f / 100.00 }
+        # Fetch product purchases
+        one_time_payments = get_one_time_payments(product_ids)
+        one_time_payments.each { |c| amount += c[:price].to_f / 100.00 }
 
-      # get number of subscribers
-      SiteSetting.discourse_subscriptions_campaign_subscribers = subscriptions&.length.to_i
+        # get number of subscribers
+        SiteSetting.discourse_subscriptions_campaign_subscribers = subscriptions&.length.to_i
 
-      # calculate amount raised
-      subscriptions&.each do |sub|
-        sub_amount = calculate_monthly_amount(sub)
-        amount += sub_amount / 100.00
-      end
+        # calculate amount raised
+        subscriptions&.each do |sub|
+          sub_amount = calculate_monthly_amount(sub)
+          amount += sub_amount / 100.00
+        end
 
-      SiteSetting.discourse_subscriptions_campaign_amount_raised = amount.round(2)
+        SiteSetting.discourse_subscriptions_campaign_amount_raised = amount.round(2)
 
-      check_goal_status
+        check_goal_status
+      rescue ::Stripe::AuthenticationError
+        message =
+          I18n.t(
+            "discourse_subscriptions.auth.invalid",
+            settingsUrl: "admin/site_settings/category/discourse_subscriptions",
+          )
+        AdminDashboardData.add_found_scheduled_check_problem(
```

Comment:
Looking at how this plugin is set up more, I see you want a JIT kind of message for this. Usually the better solution for this is to use `AdminDashboardData.add_problem_message`. Sadly though this has no plugin extensibility (nor does anything in `AdminDashboardData` atm :cry: ), since all the messages have to be defined upfront:

https://github.com/discourse/discourse/blob/1ed09234f98b68d13ed7b49c2588155c7d8c9947/app/models/admin_dashboard_data.rb#L198-L202

So I would recommend a couple of things. First of all make a commit to core to:

1. Add a filtered register in `DiscoursePluginRegistry` using `define_filtered_register`, which could be called something like `admin_problem_messages`
2. Alter `AdminDashboardData` to also iterate through `DiscoursePluginRegistry.admin_problem_messages` here https://github.com/discourse/discourse/blob/1ed09234f98b68d13ed7b49c2588155c7d8c9947/app/models/admin_dashboard_data.rb#L66-L69

Then change this PR to:

1. In plugin.rb, call `DiscoursePluginRegistry.register_admin_problem_message("discourse_subscriptions.some_i18n_key", self)
1. Call `AdminDashboardData.add_problem_message("discourse_subscriptions.some_i18n_key")`
1. Change the clear to use `AdminDashboardData.clear_problem_message("discourse_subscriptions.some_i18n_key")`

In addition, I would also add a site setting validator to validate that the Stripe key is valid when setting `discourse_subscriptions_secret_key`.

If you need some help with this, I think @Drenmi would be the person to do it.

Reply:
I know this is more more work, but it is more correct. This is why I started up a topic too to discuss the new messaging interface for admins, because like we talked about it would be good to let plugins do this kind of thing easier :)

Reply:
A few points..
1. the error is only known when the API call is made. so using core's check schedule of `10.minutes` would mean core is invoking the API every so often, even if the value is correct.
2. There is no way to "validate" the API key without an API call. In the gem, Stripe's recommended way is to use the web interface at https://stripe.com/api.

In defining this new interface on core, I will leave it to the staff exp team. With social auth validation coming up there will prob be more opinions on this and it will probably be a better use of time.

Reply:
> There is no way to "validate" the API key without an API call. In the gem, Stripe's recommended way is to use the web interface at https://stripe.com/api.

Yeah there will be a similar problem with the social auth checks. Any API call will do; doing a `GET /v1/balance` for stripe seems easiest, but people are also suggesting creating a token (but that is not very nice to have to create something).

> In defining this new interface on core, I will leave it to the staff exp team. With social auth validation coming up there will prob be more to-and-fro on this and it will probably be a better use of time. 

Yep sure makes sense to leave it to us :+1: There will definitely be plenty to do in this area to make things consistent.

Reply:
Thanks Martin!